http3/response_writer.go

```diff
@@ -81,6 +88,75 @@ func (w *responseWriter) Write(p []byte) (int, error) {
 
 func (w *responseWriter) Flush() {}
 
+func (w *responseWriter) promoteTrailer() {
+	for k, vv := range w.header {
+		if !strings.HasPrefix(k, http2.TrailerPrefix) {
+			continue
+		}
+		trailerKey := strings.TrimPrefix(k, http2.TrailerPrefix)
+		w.declareTrailer(trailerKey)
+		w.header[http.CanonicalHeaderKey(trailerKey)] = vv
+	}
+}
+
+func strSliceContains(ss []string, s string) bool {
+	for _, v := range ss {
+		if v == s {
+			return true
+		}
+	}
+	return false
+}
+
+func (w *responseWriter) declareTrailer(k string) {
+	k = http.CanonicalHeaderKey(k)
+	if !httpguts.ValidTrailerHeader(k) {
+		// Forbidden by RFC 7230, section 4.1.2.
+		w.logger.Debugf("ignoring invalid trailer %q", k)
+		return
+	}
+	if !strSliceContains(w.trailers, k) {
+		w.trailers = append(w.trailers, k)
+	}
+}
+
+func (w *responseWriter) hasNonemptyTrailers() bool {
+	for _, trailer := range w.trailers {
+		if _, ok := w.header[trailer]; ok {
+			return true
+		}
+	}
+	return false
+}
+
+func (w *responseWriter) writeTrailers() {
+	w.promoteTrailer()
+
+	if !w.hasNonemptyTrailers() {
+		return
+	}
+
+	var headers bytes.Buffer
+	enc := qpack.NewEncoder(&headers)
```

Comment:
Same as above. Should this be the same encoder?

Reply:
I'm not sure we can't because the `NewEncoder` needs a specific `io.Writer`.